### ISSUE TITLE
Binary leaking in h2_connection processes

### ISSUE THREAD

**Issue by jonasrichard** [joedevivo/chatterbox#120]
I use chatterbox as a dependency of the inaka/apns4erl project. After some hours of working I faced with ever growing process sizes (process_info(pid, memory)). After some investigation I found this:

```
process_info(pid(0,1238,0), binary).
{binary,[{139665827264464,60,1},
         {139665831210656,60,1},
         {139664600088128,60,1},
         {139665831140616,60,1},
         {139665831222376,60,1},
         {139664600244240,60,1},
         {139665827188416,60,1},
         {139664600522856,60,1},
         {139664605657712,60,1},
         {139664605954696,60,1},
         {139664605903872,60,1},
         {139664606052304,60,1},
         {139664600433752,60,1},
         {139664605649112,60,1},
         {139664531257216,60,1},
         {139665827110208,60,1},
         {139664605153864,60,1},
         {139665827411976,60,1},
         {139665827397160,60,1},
         {139665827115824,60,1},
         {139664605243344,60,1},
         {139664605133416,60,1},
         {139664605420536,60,1},
         {139664605405256,60,1},
         {139664637840368,60,...},
         {139664638044448,...},
         {...}|...]}
(myhost@my_ip)48> length(element(2, v(47))).
1851
```

Do you have a clue where this binary leak can occur?

**Comment by ferd** [comment]
If you call for a manual garbage collection of the process, do these go away? If so the leak is probably fixable through well-placed hibernation calls more than anything else.

**Comment by jonasrichard** [comment]
I did a `erlang:garbage_collect/0` but the binaries didn't go away. I don't know the process structure of chatterbox, but I think there are two processes pass the binaries, and after some matching one process hold refc binaries. It is hard to find when those binaries remain, because there were at least 50,000 http2 requests were sent.
I didn't use chatterbox directly but apns4erl.
https://github.com/inaka/apns4erl
Tomorrow I will check if other processes has something similar. Do you think in a crashdump I can see those binaries? Or the GC during crashdump will eliminate them?

**Comment by ferd** [comment]
If you can see them in the crash dump, they were really being held in memory by the process. If they are gone, that's because the VM ran a GC before writing them to the dump.

**Comment by jonasrichard** [comment]
Ok, calling the `garbage_collect(Pid)` decreased the number of refc binaries from 1932 to 84. So probably we will call GC for only those processes when the refc binaries count is more than a threashold. Does it sound right (not too odd)?

**Comment by ferd** [comment]
That can work temporarily. I think the ideal fixes in these cases for now is to modify the risky processes to run hibernation tasks from time to time when they get downtime so they force this on their own.

**Comment by jonasrichard** [comment]
It is also interesting that `process_info(memory)` and heap size in normal process_info gives me different results. And like I said, the reductions were huge. So I decided to reconnect in every 10 minute instead of the GC.

```
> process_info(P1, memory).
{memory,201303168}
> process_info(P1).
...
 {total_heap_size,17474138},
 {heap_size,2072833},
 {stack_size,10},
 {reductions,123302750526},
 {garbage_collection,[{max_heap_size,#{error_logger => true,kill => true,size => 0}},
                      {min_bin_vheap_size,46422},
                      {min_heap_size,233},
                      {fullsweep_after,65535},
                      {minor_gcs,41}]},
...
```

**Comment by tsloughter** [comment]
Looking at this again, if it makes sense we could add a `hibernate_after` option to the `gen_statem:start_link` of `h2_connection` so it hibernates if idle. 

Suggestions for what a good timeout would be?

**Comment by ferd** [comment]
Yeah the post before yours does mention a possible remediation to hibernate when idle. Someone likely just needs to go and implement it.

**Comment by sebastiw** [comment]
So we have been tracing and found out that the client responses are only cleaned up from the `h2_stream` state if the function `h2_connection:get_response/2` is called.

For instance `grpcbox_client:unary_handler` does not call this function.
for chatterbox, I think there should be an api for closing/removing a stream when not using `get_response`, and that function could then be called by client libraries such as grpcbox.

What do you think?

**Comment by tsloughter** [comment]
The grpcbox client doesn't call it because it is getting the body from messages to the client pid, right? If that is the case then it sounds more like chatterbox should have a mode that it keeps the response until it is requested and another where it only sends to some pid.

**Comment by sebastiw** [comment]
yes, there seem to be two flows in h2; one with asynchronous bangs to a notify pid, and one with some synchronous parts. might be a good idea to split them up in two, or clarify which is which.

we went with an easy quick fix which I will try to test tomorrow. Basically just added `h2_connection:get_response(Pid, Stream),` after receiving the data in `grpcbox_client:unary_handler/6`

**Comment by tsloughter** [comment]
Yea, pretty sure the notify a pid flow was added on as a quick hack at the time. Definitely could use some cleaning up.

**Comment by tsloughter** [comment]
Looking at this again.

As far as I can tell from the code the only time frames are added to the queue that becomes the response body received is if the callback is undefined. So I don't get how the response body is built up if this is the case.

**Comment by tsloughter** [comment]
Adding a hibernate_after option to h2_connection, but also merged a fix for terminating grpc streams and that may help, not sure, https://github.com/tsloughter/chatterbox/pull/15
